### PR TITLE
[LLVM] Add IRNormalizer Pass

### PR DESCRIPTION
**Pull request by justinfargnoli**
Add the llvm-canon tool. Description from the [original PR](https://reviews.llvm.org/D66029#change-wZv3yOpDdxIu):

> Added a new llvm-canon tool which aims to transform LLVM Modules into a canonical form by reordering and renaming instructions while preserving the same semantics. This tool makes it easier to spot semantic differences while diffing two modules which have undergone different transformation passes.

The current version of this tool can:

- Reorder instructions within a function.
- Rename instructions based on the operands.
- Sort commutative operands.

This code was originally written by @michalpaszkowski and [submitted to mainline LLVM](https://github.com/llvm/llvm-project/commit/14d358537f124a732adad1ec6edf3981dc9baece). However, it was quickly [reverted](https://github.com/llvm/llvm-project/commit/335de55fa3384946f1e62050f2545c0966163236) to do BuildBot errors. 

Michal presented his version of the tool in [LLVM-Canon: Shooting for Clear Diffs](https://www.youtube.com/watch?v=c9WMijSOEUg).

@AidanGoldfarb and I ported the code to the new pass manager, added more tests, and fixed some bugs related to PHI nodes that may have been the root cause of the BuildBot errors that caused the patch to be reverted. Additionally, we rewrote the implementation of instruction reordering to fix cases where the original algorithm would break use-def chains. 

Note that this is @AidanGoldfarb and I's first time submitting to LLVM. Please liberally critique the PR! 

CC @plotfi for initial review.

### CODE REVIEW

**Comment by michalpaszkowski** [comment]
Thank you @justinfargnoli and @AidanGoldfarb for taking over this work and reaching out! I will review your changes in the coming week.

CC @ChrisCummins FYI

File: llvm/lib/Transforms/Utils/IRCanonicalizer.cpp [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+/// reorderInstructions().
+///
+/// \see reorderInstructions()
+/// \param Used Pointer to the instruction whose value is used by the \p User.
+/// \param User Pointer to the instruction which uses the \p Used.
+/// \param Visited Set of visited instructions.
+void IRCanonicalizer::reorderInstruction(
+    Instruction *Used, Instruction *User,
+    SmallPtrSet<const Instruction *, 32> &Visited) {
+  if(isa<PHINode>(Used))
```
Comment (michalpaszkowski) [inline llvm/lib/Transforms/Utils/IRCanonicalizer.cpp:444]: Potentially a clang-format issue

Comment (justinfargnoli) [reply llvm/lib/Transforms/Utils/IRCanonicalizer.cpp]: Looks like clang-format has passed. Happy to change the formatting though if this isn't in line with LLVM conventions.

**Comment by AaronBallman** [comment]
Added @plotfi as a reviewer; one drive-by question is whether adding a new LLVM tool like this requires an RFC to be posted to Discourse or not (https://llvm.org/docs/DeveloperPolicy.html#introducing-new-components-into-llvm). (I don't have any opinions on the tool; just trying to help the review move along.)

**Comment by plotfi** [comment]
I recall approving the original PR on phab. I can re-review if needed. This work was approved to land before, so personally I think it might be overkill to RFC it on discourse again but I am not certain of the right process to take. 

> Added @plotfi as a reviewer; one drive-by question is whether adding a new LLVM tool like this requires an RFC to be posted to Discourse or not (https://llvm.org/docs/DeveloperPolicy.html#introducing-new-components-into-llvm). (I don't have any opinions on the tool; just trying to help the review move along.)

**Comment by nikic** [comment]
Please don't use the term canon/canonicalize for this pass or tool. LLVM has an existing notion of "canonicalization" which does coincide with what is being done here.

**Comment by AaronBallman** [comment]
> Thanks for reaching out guys.
> 
> I recall approving the original PR on phab. I can re-review if needed. This work was approved to land before, so personally I think it might be overkill to RFC it on discourse again but I am not certain of the right process to take.
> 
> > Added @plotfi as a reviewer; one drive-by question is whether adding a new LLVM tool like this requires an RFC to be posted to Discourse or not (https://llvm.org/docs/DeveloperPolicy.html#introducing-new-components-into-llvm). (I don't have any opinions on the tool; just trying to help the review move along.)

I don't have strong opinions on whether this should have an RFC or not, but the original review didn't come with an RFC that I could find either. Given that there are some concerns about what the name of the tool is, I weakly lean towards running an RFC just to make sure everyone is comfortable with the tool. But again, I don't insist, it's just a drive-by comment because I happened to notice the PR.

**Comment by plotfi** [comment]
> Please don't use the term canon/canonicalize for this pass or tool. LLVM has an existing notion of "canonicalization" which does not coincide with what is being done here.

The term comes from an older pass that did similar algorithms on MIR. They are only borrowing the term from it.

**Comment by michalpaszkowski** [comment]
> Please don't use the term canon/canonicalize for this pass or tool. LLVM has an existing notion of "canonicalization" which does not coincide with what is being done here.

@nikic I don't think the name "canon" or "canonicalizer" will lead to any confusion here. This is in fact what the pass is doing and mimics the name of MIR canonicalizer with similar assumptions as @plotfi noted . Using the term "canonicalization" for the current/existing set of passes is not really common or well defined.

As for the RFC, there was a discussion thread in the LLVM Dev Mailing: https://lists.llvm.org/pipermail/llvm-dev/2019-August/134475.html

**Comment by nikic** [comment]
Using the term for MIR is okay, because it does not have a pre-existing notion of canonical form. Doing the same for IR is not okay, because it already has a canonical form and this pass does not produce it. Calling this `ir-normalizer` instead of `ir-canonicalizer` should convey about the same intent without overloading an existing, widely used term.

**Comment by justinfargnoli** [comment]
Renamed to IRNormalizer. Requesting another round of review.

File: llvm/docs/Passes.rst [hunk 2]
```diff
@@ -543,6 +543,15 @@ variables with initializers are marked as internal.
...
 
+``ir-normalizer``: Transforms IR into a canonical form that's easier to diff
+--------------------------------------------------------
+
+This pass aims to transform LLVM Modules into a canonical form by reordering and
+renaming instructions while preserving the same semantics. The canonicalizer makes
+it easier to spot semantic differences while diffing two modules which have undergone
+two different passes.
+
+-----------------------------------
```
Comment (michalpaszkowski) [inline llvm/docs/Passes.rst:554]: Shouldn't this line be removed? and the line underscoring the "``ir-normalizer``: Transforms IR into a canonical form that's easier to diff" be made a bit longer?

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 3]
```diff
@@ -0,0 +1,637 @@
+//===--------------- IRNormalizer.cpp - IR Canonicalizer ---------------===//
```
Comment (michalpaszkowski) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:1]: Canonicalizer -> Normalizer

File: llvm/docs/Passes.rst [hunk 4]
```diff
@@ -543,6 +543,15 @@ variables with initializers are marked as internal.
 An interprocedural variant of :ref:`Sparse Conditional Constant Propagation
 <passes-sccp>`.
 
+``ir-normalizer``: Transforms IR into a canonical form that's easier to diff
+--------------------------------------------------------
+
+This pass aims to transform LLVM Modules into a canonical form by reordering and
+renaming instructions while preserving the same semantics. The canonicalizer makes
```
Comment (michalpaszkowski) [inline llvm/docs/Passes.rst:550]: canonicalizer -> normalizer

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+cl::opt<bool> IRNormalizer::FoldPreoutputs(
+    "fold-all", cl::Hidden,
+    cl::desc("Folds all regular instructions (including pre-outputs)"));
+cl::opt<bool> IRNormalizer::ReorderOperands(
+    "reorder-operands", cl::Hidden,
+    cl::desc("Sorts and reorders operands in commutative instructions"));
+
+/// Entry method to the IRNormalizer.
+///
+/// \param M Module to canonicalize.
```
Comment (michalpaszkowski) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:109]: canonicalize -> normalize

**Comment by michalpaszkowski** [comment]
I added a couple more comments regarding the name change and formatting of the docs, but apart from these the patch looks good to me.

**Comment by justinfargnoli** [comment]
Ping for another round of review.

**Review by michalpaszkowski — APPROVED** [review]
Thank you for working on this! LGTM!

File: llvm/docs/Passes.rst [hunk 6]
```diff
@@ -543,6 +543,14 @@ variables with initializers are marked as internal.
 An interprocedural variant of :ref:`Sparse Conditional Constant Propagation
 <passes-sccp>`.
 
+``ir-normalizer``: Transforms IR into a canonical form that's easier to diff
```
Comment (nikic) [inline llvm/docs/Passes.rst:546]: Looks like there are lots of leftovers of "canonical" vs "normalized".

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 7]
```diff
@@ -0,0 +1,637 @@
...
+    cl::desc("Preserves original instruction order"));
+cl::opt<bool> IRNormalizer::RenameAll(
+    "rename-all", cl::Hidden,
+    cl::desc("Renames all instructions (including user-named)"));
+cl::opt<bool> IRNormalizer::FoldPreoutputs(
+    "fold-all", cl::Hidden,
+    cl::desc("Folds all regular instructions (including pre-outputs)"));
+cl::opt<bool> IRNormalizer::ReorderOperands(
+    "reorder-operands", cl::Hidden,
+    cl::desc("Sorts and reorders operands in commutative instructions"));
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:105]: Should probably prefix these option names? They're in the global namespace.

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+
+/// Entry method to the IRNormalizer.
+///
+/// \param M Module to normalize.
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:109]: This accepts F not M.

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 9]
```diff
@@ -0,0 +1,637 @@
...
+void IRNormalizer::nameInstruction(Instruction *I) {
+  // Ensure instructions are not renamed. This is done
+  // to prevent situation where instructions are used
+  // before their definition (in phi nodes)
+  if (NamedInstructions.contains(I))
+    return;
+  NamedInstructions.insert(I);
+  // Determine the type of instruction to name.
+  if (isInitialInstruction(I)) {
+    // This is an initial instruction.
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:185]: This comment and the one below just repeat the method names :)

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+  for (auto &OP : I->operands()) {
+    if (!isa<Function>(OP)) {
+      std::string TextRepresentation;
+      raw_string_ostream Stream(TextRepresentation);
+      OP->printAsOperand(Stream, false);
+      Operands.push_back(StringRef(Stream.str()));
+    }
+  }
+
+  if (I->isCommutative())
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:223]: isCommutative() only applies to the first two operands, not all of them.

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 11]
```diff
@@ -0,0 +1,637 @@
...
+
+  // Base instruction name.
+  SmallString<256> Name;
+  Name.append("vl" + std::to_string(Hash).substr(0, 5));
+
+  // In case of CallInst, consider callee in the instruction name.
+  if (const auto *CI = dyn_cast<CallInst>(I)) {
+    Function *F = CI->getCalledFunction();
+
+    if (F != nullptr) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:248]: Braces

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 12]
```diff
@@ -0,0 +1,637 @@
...
+  // operands must be named first in the use-def
+  // walk.
+
+  // Collect operands.
+  for (auto &OP : I->operands()) {
+    if (auto *IOP = dyn_cast<Instruction>(OP)) {
+      // Walk down the use-def chain.
+      nameInstruction(IOP);
+      Operands.push_back(IOP->getName());
+    } else if (isa<Value>(OP) && !isa<Function>(OP)) {
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:297]: `isa<Value>` is redundant.

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 13]
```diff
@@ -0,0 +1,637 @@
...
+      // This must be an immediate value.
+      std::string TextRepresentation;
+      raw_string_ostream Stream(TextRepresentation);
+      OP->printAsOperand(Stream, false);
+      Operands.push_back(StringRef(Stream.str()));
+    }
+  }
+
+  if (I->isCommutative())
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:306]: Same here, only first two are commutative.

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 14]
```diff
@@ -0,0 +1,637 @@
...
+
+  // Operand opcodes for further sorting (commutative).
+  SmallVector<int, 4> OperandsOpcodes;
+
+  // Collect operand opcodes for hashing.
+  for (auto &OP : I->operands())
+    if (auto *IOP = dyn_cast<Instruction>(OP))
+      OperandsOpcodes.push_back(IOP->getOpcode());
+
+  if (I->isCommutative())
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:323]: Same here.

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp [hunk 15]
```diff
@@ -0,0 +1,637 @@
...
+    Used->moveBefore(&Used->getParent()->back());
+  }
+
+  for (auto &OP : Used->operands()) {
+    if (auto *IOP = dyn_cast<Instruction>(OP)) {
+      // Walk up the def-use tree.
+      reorderInstruction(IOP, Used, Visited);
+    }
+  }
+}
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/IRNormalizer.cpp:464]: Am I missing something, or can this reorder side-effecting instructions?

**Review by nikic — COMMENTED** [review]
Could you please update the patch description with some information on what kind of normalization the pass does?

It would also be great to post some examples, because I don't really get what kind of renaming and reordering this does just looking at the test coverage.

One bit that's particularly non-obvious to me is why this normalization has to involve hashes? That seems like it would make things more unstable and hard to understand.

I would recommend trying to run this pass (with all options enabled) over all existing tests to make sure that it doesn't cause any crashes/verifier failures. You're moving instructions around, and it's easy to get that wrong when invoke, callbr or catchswitch are involved.

I would also try to add this pass at the end of the clang pipeline and run llvm-test-suite to verify that the normalization this does is indeed semantics-preserving.

(Review notes are just nits, I haven't tried to understand the main logic yet.)

**Comment by michalpaszkowski** [comment]
[Here are the slides](https://llvm.org/devmtg/2019-10/slides/Paszkowski-LLVMCanon.pdf) that were presented at the LLVM Dev Meeting in 2019. It might be useful to include the link in the commit message for reference.

@nikic Thanks for catching more things in the review! Especially the commutative ordering of operands/arguments.

When it comes to including hashes in the naming algorithm (renaming is controlled using an option for already named values), the ideas is that by just looking at an instruction further down (a use) you can tell whether any inputs are different. And the other way around, by looking at initial instructions you can see how their "output footprint" has changed after transformations.

The example below is a bit outdated, but reflects the general idea we had when considering naming algorithms:
image
image

File: llvm/lib/Transforms/Utils/IRNormalizer.cpp (hunk 15 above)
Comment (justinfargnoli) [reply llvm/lib/Transforms/Utils/IRNormalizer.cpp:490]: Yes, it can reorder side-effecting instructions. In the process of rewriting this code.

**Comment by justinfargnoli** [comment]
> I would recommend trying to run this pass (with all options enabled) over all existing tests to make sure that it doesn't cause any crashes/verifier failures. You're moving instructions around, and it's easy to get that wrong when invoke, callbr or catchswitch are involved.

Thanks for this suggestion! It helped iron out a lot of bugs in the implementation. 

> I would also try to add this pass at the end of the clang pipeline and run llvm-test-suite to verify that the normalization this does is indeed semantics-preserving.

I'm running into some linking errors when trying to build [`llvm-test-suite`](https://github.com/llvm/llvm-test-suite/tree/main). Pinging for another round of review in the meantime.

**Comment by justinfargnoli** [comment]
Pinging @nikic for another round of review.

**Comment by plotfi** [comment]
> Pinging @nikic for another round of review.

 @nikic is there more feedback or can this PR be merged?

**Comment by justinfargnoli** [comment]
Unless other comments brought up, I'll plan on merging this on Monday the 4th.

**Comment by nikic** [comment]
> > I would also try to add this pass at the end of the clang pipeline and run llvm-test-suite to verify that the normalization this does is indeed semantics-preserving.
> 
> I'm running into some linking errors when trying to build [`llvm-test-suite`](https://github.com/llvm/llvm-test-suite/tree/main). Pinging for another round of review in the meantime.

I just tried this myself, and received the following failures. There were also a bunch of hanging processes I had to kill, but I didn't note down which those were.

```
  test-suite :: MicroBenchmarks/LCALS/SubsetALambdaLoops/lcalsALambda.test
  test-suite :: MicroBenchmarks/LCALS/SubsetARawLoops/lcalsARaw.test
  test-suite :: MicroBenchmarks/LCALS/SubsetBLambdaLoops/lcalsBLambda.test
  test-suite :: MicroBenchmarks/LCALS/SubsetBRawLoops/lcalsBRaw.test
  test-suite :: MicroBenchmarks/LCALS/SubsetCLambdaLoops/lcalsCLambda.test
  test-suite :: MicroBenchmarks/LCALS/SubsetCRawLoops/lcalsCRaw.test
  test-suite :: MultiSource/Applications/lua/lua.test
  test-suite :: SingleSource/Benchmarks/CoyoteBench/huffbench.test
```

So I think there are still correctness issues in this pass.

File: llvm/test/Transforms/IRNormalizer/reordering-instructions.ll [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+; CHECK: %b
+; CHECK: %d
+; CHECK: %a
+; CHECK: %c
+  %a = fmul double %a0, %a1
+  %b = fmul double %a0, 2.000000e+00
+  %c = fmul double %a, 6.000000e+00
+  %d = fmul double %b, 6.000000e+00
+  ret double %d
+}
```
Comment (nikic) [inline llvm/test/Transforms/IRNormalizer/reordering-instructions.ll:14]: I'd suggest to add some more tests for this, in particular for instructions with side-effects or accessing memory.

If possible (i.e. for tests that don't include hashes in the output) you should use update_test_checks to generate the CHECK lines.

**Comment by justinfargnoli** [comment]
TODO: Sort function order

**Comment by dtcxzyw** [comment]
Any progress? It would be helpful for me to avoid reviewing tons of diff in https://github.com/dtcxzyw/llvm-opt-benchmark/pulls :)

Can you resolve merge conflicts first?

**Comment by justinfargnoli** [comment]
Hi @dtcxzyw, thanks for reaching out! 

This MR is blocked because I cannot reproduce the failures @nikic found [here](https://github.com/llvm/llvm-project/pull/68176#issuecomment-1836365664). (See [this comment](https://github.com/llvm/llvm-project/pull/68176#issuecomment-1818308163) to understand the problem I'm running into.)  

The other things that I've mentioned should be easy to fix. 

If you (or anyone else who's interested in helping) can reproduce/debug the failure @nikic found, that'd be a big help toward getting this PR over the line.

**Comment by dtcxzyw** [comment]
> Hi @dtcxzyw, thanks for reaching out!
> 
> This MR is blocked because I cannot reproduce the failures @nikic found [here](https://github.com/llvm/llvm-project/pull/68176#issuecomment-1836365664). (See [this comment](https://github.com/llvm/llvm-project/pull/68176#issuecomment-1818308163) to understand the problem I'm running into.)
> 
> The other things that I've mentioned should be easy to fix.
> 
> If you (or anyone else who's interested in helping) can reproduce/debug the failure @nikic found, that'd be a big help toward getting this PR over the line.

Test result on my [dataset](https://github.com/dtcxzyw/llvm-opt-benchmark):
```
llvm-opt-benchmark/bench/ruby/optimized/vm.ll Crash
llvm-opt-benchmark/bench/ruby/optimized/regexec.ll Crash
llvm-opt-benchmark/bench/oniguruma/optimized/regexec.ll Crash
llvm-opt-benchmark/bench/jq/optimized/regexec.ll Crash
llvm-opt-benchmark/bench/rust-analyzer-rs/optimized/12c5ozyvkyoo7zj1.ll Timeout
llvm-opt-benchmark/bench/rust-analyzer-rs/optimized/11fzjzs7q1iz2s6j.ll Timeout
llvm-opt-benchmark/bench/rust-analyzer-rs/optimized/1ritseoble5k1o6v.ll Timeout
llvm-opt-benchmark/bench/rust-analyzer-rs/optimized/3mo33cygxsz2niby.ll Timeout
llvm-opt-benchmark/bench/rust-analyzer-rs/optimized/2oxkgole77u9464n.ll Timeout
llvm-opt-benchmark/bench/rust-analyzer-rs/optimized/57yy3hfi5uyj25tj.ll Timeout
llvm-opt-benchmark/bench/quickjs/optimized/quickjs.ll Crash
llvm-opt-benchmark/bench/spike/optimized/processor.ll Timeout
llvm-opt-benchmark/bench/duckdb/optimized/is_distinct_from.cpp.ll Timeout
llvm-opt-benchmark/bench/cvc5/optimized/options_public.cpp.ll Crash
llvm-opt-benchmark/bench/wireshark/optimized/enterprises.c.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/4h15go5rbmdr09f8.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/y59h8jnu9k73im6.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/59na08zlxqp9p17s.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/40iecfzh8kpu4oof.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/cahtbyi0e0azi7k.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/1d1qxhuj8x85dazd.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/2jdafynperrjfnwp.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/2ugp26prskc4lvz4.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/5ayhis8cfs1k9t14.ll Timeout
llvm-opt-benchmark/bench/glog/optimized/symbolize_unittest.cc.ll Crash
llvm-opt-benchmark/bench/ockam-rs/optimized/scbbgml6cvr1nwt.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/1j8f46ag92qmaepd.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/2qz3huiqqut486jt.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/4df8gyzy0u3roc94.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/39elqh75xhewanjo.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/445trp5tofqtvmh3.ll Timeout
llvm-opt-benchmark/bench/ockam-rs/optimized/lcclztcuaxx9mqt.ll Timeout
llvm-opt-benchmark/bench/faiss/optimized/pq4_fast_scan_search_qbs.cpp.ll Timeout
llvm-opt-benchmark/bench/meshlab/optimized/filter_mesh_alpha_wrap.cpp.ll Crash
llvm-opt-benchmark/bench/meshlab/optimized/filter_mesh_booleans.cpp.ll Crash
llvm-opt-benchmark/bench/typst-rs/optimized/26ex1w19ixz6ifuk.ll Timeout
llvm-opt-benchmark/bench/typst-rs/optimized/22a0ijn9jg8ilzgs.ll Timeout
llvm-opt-benchmark/bench/typst-rs/optimized/1ru1rhojhbz2vfey.ll Timeout
llvm-opt-benchmark/bench/arrow/optimized/builder.cc.ll Timeout
llvm-opt-benchmark/bench/delta-rs/optimized/3s3f23ex69abms10.ll Timeout
```

**Comment by dtcxzyw** [comment]
> > > I would also try to add this pass at the end of the clang pipeline and run llvm-test-suite to verify that the normalization this does is indeed semantics-preserving.
> > 
> > 
> > I'm running into some linking errors when trying to build [`llvm-test-suite`](https://github.com/llvm/llvm-test-suite/tree/main). Pinging for another round of review in the meantime.
> 
> I just tried this myself, and received the following failures. There were also a bunch of hanging processes I had to kill, but I didn't note down which those were.
> 
> ```
>   test-suite :: MicroBenchmarks/LCALS/SubsetALambdaLoops/lcalsALambda.test
>   test-suite :: MicroBenchmarks/LCALS/SubsetARawLoops/lcalsARaw.test
>   test-suite :: MicroBenchmarks/LCALS/SubsetBLambdaLoops/lcalsBLambda.test
>   test-suite :: MicroBenchmarks/LCALS/SubsetBRawLoops/lcalsBRaw.test
>   test-suite :: MicroBenchmarks/LCALS/SubsetCLambdaLoops/lcalsCLambda.test
>   test-suite :: MicroBenchmarks/LCALS/SubsetCRawLoops/lcalsCRaw.test
>   test-suite :: MultiSource/Applications/lua/lua.test
>   test-suite :: SingleSource/Benchmarks/CoyoteBench/huffbench.test
> ```
> 
> So I think there are still correctness issues in this pass.

I cannot reproduce these failures on my x86 machine.
```
Testing Time: 298.93s

Total Discovered Tests: 2438
  Passed: 2438 (100.00%)
```

**Comment by dtcxzyw** [comment]
> Any progress? It would be helpful for me to avoid reviewing tons of diff in https://github.com/dtcxzyw/llvm-opt-benchmark/pulls :)
> 
> Can you resolve merge conflicts first?

Emm, as this pass is very time-consuming (even longer than -O3), I plan to create a simpler one for personal use. Anyway, thanks for your idea!

**Comment by plotfi** [comment]
@justinfargnoli Is there anything I can do to help with moving this forward, what seems to be the blocker on this at this point (asking to re-attain context)?

**Comment by justinfargnoli** [comment]
Hi @plotfi, thanks for pinging me on this! Generally speaking, two things deserve additional focus:

- UX testing to find:
  - Improvements to existing features 
  -  New features to add
- Additional testing to ensure that the pass preserves the semantics of the IR. 
  - e.g. I want to add this pass to the `nvcc` optimization pipeline and run it on NVIDIA's internal test suite. 
  - If Meta has similar test suites, it would be great to see if this pass introduces any functional issues on it. 

---

> I cannot reproduce these failures on my x86 machine.

Big thanks to @dtcxzyw for running the test-suite. Since @dtcxzyw did not find any test failures, this MR is ready for another round of review. 

If things look good to @plotfi and @nikic, then we should be good to get this merged!

**Comment by justinfargnoli** [comment]
Ping @plotfi @nikic for review.

**Comment by justinfargnoli** [comment]
Ping @plotfi @nikic for review.

**Comment by plotfi** [comment]
LGTM, I would prefer if we landed this and added improvements as needed

**Comment by vvereschaka** [comment]
Looks like these failed tests on [llvm-clang-x86_64-expensive-checks-ubuntu](https://lab.llvm.org/buildbot/#/builders/187) builder are related with this PR:
```
LLVM ERROR: Function @test changed by IRNormalizerPass without invalidating analyses
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/build/bin/opt -S -passes=normalize
1.	Running pass "function(normalize)" on module "<stdin>"
2.	Running pass "normalize" on function "test"
 #0 0x00005643d9c9fdaa llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:22
...
```
https://lab.llvm.org/buildbot/#/builders/187/builds/2031

**Comment by justinfargnoli** [comment]
@vvereschaka, I agree. 

Note: The problem is that `IRNormalizer` produces a change that invalidates an analysis. It's not that IRNormalizer should declare that it's not preserving an analysis. 

I likely won't be able to fix this for a few days. @vvereschaka, should I revert 1295d2e6da2fe90f3b770ab1d35bf5caecd38bed in the meantime?

**Comment by vvereschaka** [comment]
Reverting the changes is the optimal way in those cases. Yes, would you revert this PR. Thank you.

**Pull request merged**